lib/modules.nix

```diff
@@ -490,7 +502,15 @@ rec {
               firstOption = findFirst (m: isOption m.options) "" decls;
               firstNonOption = findFirst (m: !isOption m.options) "" decls;
             in
-              throw "The option `${showOption loc}' in `${firstOption._file}' is a prefix of options in `${firstNonOption._file}'."
+              if firstOption.options.type?isSubmodule && firstOption.options.type.isSubmodule
+              then
+                let opt = fixupOptionType loc (mergeOptionDecls loc (map optionTreeToOption decls));
+                in {
+                  matchedOptions = evalOptionValue loc opt defns';
+                  unmatchedDefns = [];
+                }
```

Comment:
This could probably be deduplicated with the code just above.

Reply:
Done.
It could be made even less repetitive, but that would require more conditionals or function calls that I'd prefer not to add because of evaluation performance.

Reply:
I think what I meant was that the first two cases can be combined into one as they return the same value (`map optionTreeToOption decls == decls` in the first case).

```nix
if optionDecls == [] then
  mergeModules' loc decls defns
else if (lib.head optionDecls).options.type.name != "submodule" then
  throw "The option ... is a prefix of ..."
else
  let opt = fixupOptionType loc (mergeOptionDecls loc (map optionTreeToOption decls));
  in {
    matchedOptions = evalOptionValue loc opt defns';
    unmatchedDefns = [];
  }
```

I don't think this hurts performance? It might slightly change the semantics if `decls == []`, but is that case even possible?

Reply:
> I guess it calls optionTreeToOption more.

Exactly. Calls are quite expensive.

> make `map optionTreeToOption` a fold

I don't expect going from two list function to a single currying list function to be an improvement. Having to then store both results as an intermediate value is very likely to make it more expensive.

> Does this kind of micro-optimisation really matter in practice?

Often it doesn't, but this code is called very frequently. OfBorg's total allocated bytes are down by 1/800th. I'd say that's not insignificant. I'd speculate based on the numbers that a `builtins.count` could achieve four times that, or 0.5%.

I don't think we need to improve performance any further in this PR, but It will be interesting to try something in a follow up, so we can see the effect of a single change.